src/shogun/preprocessor/RandomFourierGaussPreproc.cpp

```diff
@@ -140,264 +35,133 @@ EFeatureType RandomFourierGaussPreproc::get_feature_type() {
-	{
-	randomcoeff_additive=SG_MALLOC(float64_t, cur_dim_feature_space);
-	randomcoeff_multiplicative=SG_MALLOC(float64_t, cur_dim_feature_space*cur_dim_input_space);
-
-	std::copy(randomcoeff_additive2, randomcoeff_additive2
-			+ dim_feature_space, randomcoeff_additive);
-	std::copy(randomcoeff_multiplicative2, randomcoeff_multiplicative2
-			+ cur_dim_feature_space*cur_dim_input_space, randomcoeff_multiplicative);
-	}
+	m_dims = other.m_dims;
+	m_kernelwidth = other.m_kernelwidth;
 
+	m_basis = other.m_basis;
+	m_offset = other.m_offset;
 }
 
 void RandomFourierGaussPreproc::fit(std::shared_ptr<Features> f)
 {
-	if (dim_feature_space <= 0) {
-		throw ShogunException(
-				"CRandomFourierGaussPreproc::init (Features *f): dim_feature_space<=0 is not allowed, use void set_dim_feature_space(const int32 dim) before!\n");
-	}
+	require(m_dims > 0, "Feature space dimension must be greater than zero");
+	require(m_kernelwidth > 0, "Kernelwidth must be greater than zero");
 
-	io::info("calling CRandomFourierGaussPreproc::init(...)");
 	int32_t num_features =
 	    f->as<DenseFeatures<float64_t>>()->get_num_features();
 
-	if (!test_rfinited()) {
-		dim_input_space = num_features;
-		init_randomcoefficients();
-		ASSERT( test_rfinited())
-	} else {
-		dim_input_space = num_features;
-		// does not reinit if dimension is the same to avoid overriding a previous call of set_randomcoefficients(...)
-		init_randomcoefficients();
-	}
+	require(
+	    num_features > 0, "Input space dimension must be greater than zero");
+
+	io::info("calling CRandomFourierGaussPreproc::init(...)");
+
+	generate_coeff(num_features);
 }
 
-SGVector<float64_t> RandomFourierGaussPreproc::apply_to_feature_vector(SGVector<float64_t> vector)
+SGVector<float64_t> RandomFourierGaussPreproc::apply_to_feature_vector(const SGVector<float64_t>& vector)
 {
-	if (!test_rfinited()) {
-		throw ShogunException(
-				"float64_t * CRandomFourierGaussPreproc::apply_to_feature_vector(...): test_rfinited()==false: you need to call before CRandomFourierGaussPreproc::init (Features *f) OR	1. set_dim_feature_space(const int32 dim), 2. set_dim_input_space(const int32 dim), 3. init_randomcoefficients() or set_randomcoefficients(...) \n");
-	}
+	SGVector<float64_t> res(m_dims);
+	SGVector<float64_t> offset_vector(m_dims);
 
-	float64_t val = std::sqrt(2.0 / cur_dim_feature_space);
-	SGVector<float64_t> res(cur_dim_feature_space);
+	if(m_basis.size() == 0 || m_offset.size() == 0)
+		error("Random coefficients not instantiated, use fit() before");
 
-	for (int32_t od = 0; od < cur_dim_feature_space; ++od) {
-		SGVector<float64_t> wrapper(randomcoeff_multiplicative+od*cur_dim_input_space, cur_dim_input_space, false);
-		res[od] = val * cos(randomcoeff_additive[od] + linalg::dot(vector, wrapper));
-	}
+	linalg::add(m_offset,linalg::matrix_prod(m_basis,vector),offset_vector);
+
+	for(auto idx : range(m_dims))
+		res[idx] = std::sqrt(2.0 / m_dims) * cos(offset_vector[idx]);
```

Comment:
once the cos is in linalg, this can be vectorized as well right? :)

Reply:
HAha ofcourse!!. It looks a bit ugly though, the "linalg::" gets chained up. Anyways better when you see it for yourself  and maybe we can go back and forth on that ;)